src/spv/bitcoin/BRPeer.cpp

```diff
@@ -833,142 +826,37 @@ static int _BRPeerAcceptMessage(BRPeer *peer, const uint8_t *msg, size_t msgLen,
-        setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (sockopt_arg_type) &on, sizeof(on));
-#ifdef SO_NOSIGPIPE // BSD based systems have a SO_NOSIGPIPE socket option to supress SIGPIPE signals
-        setsockopt(sock, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on));
-#endif
-#endif
-
-#ifdef WIN32
-        u_long nBlock = 1; // Non-Blocking
-        if (ioctlsocket(sock, FIONBIO, &nBlock) == SOCKET_ERROR) {
-            r = 0;
-        }
-#else
-        arg = fcntl(sock, F_GETFL, NULL);
-        if (arg < 0 || fcntl(sock, F_SETFL, arg | O_NONBLOCK) == SOCKET_ERROR) r = 0; // temporarily set socket non-blocking
-#endif
-        if (! r) err = WSAGetLastError();
-    }
-
-    if (r) {
-        memset(&addr, 0, sizeof(addr));
-        
-        if (domain == PF_INET6) {
-            ((struct sockaddr_in6 *)&addr)->sin6_family = AF_INET6;
-            ((struct sockaddr_in6 *)&addr)->sin6_addr = *(struct in6_addr *)&peer->address;
-            ((struct sockaddr_in6 *)&addr)->sin6_port = htons(peer->port);
-            addrLen = sizeof(struct sockaddr_in6);
-        }
-        else {
-            ((struct sockaddr_in *)&addr)->sin_family = AF_INET;
-            ((struct sockaddr_in *)&addr)->sin_addr = *(struct in_addr *)&peer->address.u32[3];
-            ((struct sockaddr_in *)&addr)->sin_port = htons(peer->port);
-            addrLen = sizeof(struct sockaddr_in);
-        }
-
-#ifdef WIN32
-        if (connect(sock, (struct sockaddr *)&addr, addrLen) == SOCKET_ERROR) err = WSAGetLastError();
-#else
-        if (connect(sock, (struct sockaddr *)&addr, addrLen) < 0) err = errno;
-#endif
-
-        // Connection in progress.
-#ifdef WIN32
-        if (err == WSAEINPROGRESS || err == WSAEWOULDBLOCK || err == WSAEINVAL) {
-#else
-        if (err == EINPROGRESS) {
-#endif
-            err = 0;
-            optLen = sizeof(err);
-            tv.tv_sec = timeout;
-            tv.tv_usec = (long)(timeout*1000000) % 1000000;
-            FD_ZERO(&fds);
-            FD_SET(sock, &fds);
-            count = select(sock + 1, NULL, &fds, NULL, &tv);
-
-            if (count <= 0 || getsockopt(sock, SOL_SOCKET, SO_ERROR, (sockopt_arg_type)&err, &optLen) == SOCKET_ERROR || err) {
-                if (count == 0) err = ETIMEDOUT;
-                if (count < 0 || ! err) err = WSAGetLastError();
-                r = 0;
-            }
-        }
-        else if (err) {
-            r = 0;
-        }
-
-        if (r) peer_log(peer, "socket connected");
-
-#ifdef WIN32
-    u_long nZero = 0;
-    ioctlsocket(sock, FIONBIO, &nZero);
-#else
-    fcntl(sock, F_SETFL, arg);
-#endif
-    }
-    if (! r && err) peer_log(peer, "connect error: %d: %s", err, strerror(err));
-    if (error && err) *error = err;
-    return r;
-}
-
 static int _peerCheckAndGetSocket (BRPeerContext *ctx, SOCKET *socket) {
     int exists;
 
     ctx->lock.lock();
-    exists = ctx->socket != INVALID_SOCKET;
+    exists = ctx->socket >= 0;
```

Comment:
On Windows, invalid socket is 0, while on other platforms, it's >=0.

Reply:
Since windows does not real use fds, rather kernel objects and handles. Believe 0 is error and anything non zero is valid.

Reply:
Looks like Bitcoin code, the hypothesis I'm forming from a quick glance is that, it assumes that Linux fds for these sockets will never be 0 (since that's stdin) and as such uses Window's 0 as the SOCKET_INVALID, and then on each failure, sets it manually to SOCKET_INVALID to ensure it's consistent. Perhaps, we should do the same?